lib/fluent/plugin/parser.rb

```diff
@@ -65,12 +111,43 @@ def configure(conf)
         @null_value_regexp = @null_value_pattern && Regexp.new(@null_value_pattern)
         @type_converters = build_type_converters(@types)
         @execute_convert_values = @type_converters || @null_value_regexp || @null_empty_string
+        @timeout_checker = if @timeout
+                             class << self
+                               alias_method :parse_orig, :parse
+                               alias_method :parse, :parse_with_timeout
+                             end
+                             TimeoutChecker.new(@timeout)
+                           else
+                             nil
+                           end
+      end
+
+      def start
+        super
+
+        @timeout_checker.start if @timeout_checker
+      end
+
+      def stop
+        super
+
+        @timeout_checker.stop if @timeout_checker
       end
 
       def parse(text, &block)
         raise NotImplementedError, "Implement this method in child class"
       end
 
+      def parse_with_timeout(text, &block)
+        @timeout_checker.execute {
```

Comment:
Why don't you use `Timeout` class? 

```rb
def parse_with_timeout(text, &block)
  Timeout.timeout do
    parse_orign(text, &block)
  end
rescue TimeoutError
  log.warn "parsing timed out with #{self.class}: text = #{text}"
  # Return nil instead of raising error. in_tail or other plugin can emit broken line.
  yield nil, nil
end

```

Reply:
`Timeout` launches one thread per action. This is not good parser.
When receive 1000 events, 1000 threads are launched.

Reply:
> Timeout launches one thread per action. 

I see. then, this implementation is better.